vllm/core/block_manager_v2.py

```diff
@@ -274,6 +285,50 @@ def mark_blocks_as_computed(self, seq_group: SequenceGroup):
         # So this function is useless for block_v2.
         pass
 
+    def _get_and_update_computed_block_ids(self, seqs):
+        """Handles caching of per-sequence computed block ids. 
+        When a sequence appears for the first time, it traverses all of the 
+        blocks and detects the prefix of blocks that is computed. On the
+        subsequent times, it only traverses the new blocks that were added 
+        and updates the already recorded prefix of blocks with the newly 
+        computed blocks.
+        """
+        ret = []
+        for seq in seqs:
+            seq_id = seq.seq_id
+
+            # Get block ids of this sequence, while not considering the
+            # last block
+            block_ids = self.block_tables[seq_id].physical_block_ids[:-1]
+
+            # Here we cache the detection of computed_block_ids for seq_id.
+            # Since computed_block_ids form a prefix of block_ids,
+            # the first time we see seq_id, we detect computed_block_ids
+            # fully and store them in the cache. In the next times we see
+            # seq_id, we detect computed_block_ids incrementally, by looking
+            # only at the new blocks that come after the cached
+            # computed_block_ids
+            if seq_id not in self._cached_computed_seq_blocks:
+                # First time init for seq_id => Detect fully
+                computed_block_ids = self.block_allocator.get_computed_block_ids(  # noqa: E501
+                    [], block_ids)
+                self._cached_computed_seq_blocks[seq_id] = computed_block_ids
+            else:
+                computed_block_ids = self._cached_computed_seq_blocks[seq_id]
+                if len(computed_block_ids) < len(block_ids):
+                    # Incremental init for seq_id => Look only at the new blocks
+                    computed_block_ids = self.block_allocator.get_computed_block_ids(  # noqa: E501
+                        computed_block_ids, block_ids)
+                    self._cached_computed_seq_blocks[
+                        seq_id] = computed_block_ids
+                else:
+                    # Cache HIT
+                    assert len(computed_block_ids) == len(block_ids)
```

Comment:
This will still result in constant recomputation in the worst case. I think we can do the following:
1. After the first run, if `len(computed_block_ids) != len(block_ids)`, we know that we will never add any extra blocks to `computed_block_ids` (since we'd have a gap otherwise). Therefore, we should save that as a boolean in the cache alongside the computed block ids
2. In the subsequent runs, if the `seq_id` is present in cache, but the boolean is False, we just return the cached computed block ids without calling `get_computed_block_ids`. Otherwise, if the boolean is true, we call `get_computed_block_ids` for the new blocks and save in cache, with the `len(computed_block_ids) == len(block_ids)` boolean.

let me know if this makes sense? I may be missing something here.

Reply:
Here's the suggested change:
```python
    def _get_and_update_computed_block_ids(self, seqs):
        """Handles caching of per-sequence computed block ids. 
        When a sequence appears for the first time, it traverses all of the 
        blocks and detects the prefix of blocks that is computed. On the
        subsequent times, it only traverses the new blocks that were added 
        and updates the already recorded prefix of blocks with the newly 
        computed blocks.
        """
        ret = []
        for seq in seqs:
            seq_id = seq.seq_id

            # Get block ids of this sequence, while not considering the
            # last block
            block_ids = self.block_tables[seq_id].physical_block_ids[:-1]

            # Here we cache the detection of computed_block_ids for seq_id.
            # Since computed_block_ids form a prefix of block_ids,
            # the first time we see seq_id, we detect computed_block_ids
            # fully and store them in the cache. In the next times we see
            # seq_id, we detect computed_block_ids incrementally, by looking
            # only at the new blocks that come after the cached
            # computed_block_ids
            if seq_id not in self._cached_computed_seq_blocks:
                # First time init for seq_id => Detect fully
                computed_block_ids = self.block_allocator.get_computed_block_ids(  # noqa: E501
                    [], block_ids)
                self._cached_computed_seq_blocks[seq_id] = (computed_block_ids, len(block_ids)==len(computed_block_ids))
            else:
                computed_block_ids, should_continue_adding = self._cached_computed_seq_blocks[seq_id]
                if should_continue_adding:
                    if len(computed_block_ids) < len(block_ids):
                        # Incremental init for seq_id => Look only at the new blocks
                        computed_block_ids = self.block_allocator.get_computed_block_ids(  # noqa: E501
                            computed_block_ids, block_ids)
                        self._cached_computed_seq_blocks[
                            seq_id] = (computed_block_ids, len(block_ids)==len(computed_block_ids))
                    else:
                        # Cache HIT
                        assert len(computed_block_ids) == len(block_ids)

            ret.append(computed_block_ids)

        return ret
```

Reply:
@Yard1 and I discussed this in more detail and this is a really good suggestion that should help with performance. Will add this to the algorithm.

Reply:
@Yard1 Added your idea inside. All works.